### PR TITLE
Fix issue #1564, add seccomp limit syscall argument

### PR DESCRIPTION
**Pull request by lifeng68**
Fix issue #1564 
To support limit syscall arguments, extend the version 2 file to the following format:

```
syscall_name action [index,value,op,valueTwo] [index,value,op]...
```

for one arguments, [index,value,valueTwo,op]

index: the index for syscall arguments（type uint）

value: the value for syscall arguments （type uint64）

op: the operator for syscall arguments(string), a valid list of constants as of libseccomp v2.3.2 is
SCMP_CMP_NE,SCMP_CMP_LE,SCMP_CMP_LE, SCMP_CMP_EQ, SCMP_CMP_GE,
SCMP_CMP_GT, SCMP_CMP_MASKED_EQ, or !=,<=,==,>=,>,&=

valueTwo: the value for syscall arguments only used for mask eq（type uint64, optional）

For example:

```
2
blacklist allow
reject_force_umount  # comment this to allow umount -f;  not recommended
[all]
kexec_load errno 1 [0,1,SCMP_CMP_LE][3,1,==][5,1,SCMP_CMP_MASKED_EQ,1]
open_by_handle_at errno 1
init_module errno 1
finit_module errno 1
delete_module errno 1
```
Signed-off-by: LiFeng <lifeng68@huawei.com>

### CODE REVIEW

**Comment by brauner** [comment]
`clang`'s being a good compiler as always:)

```
../../../src/lxc/seccomp.c:214:9: error: comparison of constant -1 with
      expression of type 'enum scmp_compare' is always false
      [-Werror,-Wtautological-constant-out-of-range-compare]
        if (op == -1) {
            ~~ ^  ~~
```

**Comment by brauner** [comment]
This is an implementation based on the suggestion by @0x0916 I take it?
I saw that @hallyn and @Blub thought it might be a good idea to directly have the ebpf listed in there if we extend the interface now. So I wait for them to comment. At a first glance the approach doesn't look unreasonable to me. I'd like to hear some more arguments though why the EBPF approach makes more sense.

**Comment by Blub** [comment]
On the one hand I still think bpf asm (numeric `struct bpf_insn` lines) would be the most generic way, on the other hand it might scare people off...

This does look like a usable compromise. One thing I'd definitely change though is the `args:num` - this can simply be dropped. Just parse to the end of the line and count the conditions.

As for the rest, it would probably be nice if `parse_v2_rule_op` would also map `==`, `!=`, `<=`, etc.
It would probably also be easier if the mask was last and optional. `[0,3,==], [1,4,==,7]` would be a non-masked and a masked comparison. (I'm deliberately not suggesting putting the operator in the middle because `[0,==,3]` or worse `[0 == 3]` would simply give the wrong impression...)

Some nitpicking: `%llu` specifically parses an unsigned _decimal_ integer whereas `%lli` supports hex if it starts with `0x`, so that's worth considering as well (and negative numbers may make sense anyway).

**Comment by brauner** [comment]
> On the one hand I still think bpf asm (numeric struct bpf_insn lines) would be the most generic way, on the other hand it might scare people off...

Yeah, I agree but let's see what @hallyn thinks.

> This does look like a usable compromise. One thing I'd definitely change though is the args:num - this can simply be dropped. Just parse to the end of the line and count the conditions.

Yeah, I second that. It would be good if we could just generically parse this without relying on the `args:num` fields.

As for the rest, it would probably be nice if parse_v2_rule_op would also map ==, !=, <=, etc.
It would probably also be easier if the mask was last and optional. [0,3,==], [1,4,==,7] would be a non-masked and a masked comparison. (I'm deliberately not suggesting putting the operator in the middle because [0,==,3] or worse [0 == 3] would simply give the wrong impression...)

Agreed as well.

**Comment by hallyn** [comment]
What exactly should happen with:

```
kexec_load errno 1 args:3 [0,1,0,SCMP_CMP_LE][3,1,0,SCMP_CMP_GT][5,1,0,SCMP_CMP_MASKED_EQ]
```
?

`errno 1` is the *default* action;  then there are some conditions...  but what happens when those conditions are met?

valueTwo is only ever used for the masked eq, is that right?

How about

```
rule := syscall [condition_or_rule] ...
condition_or_rule := [comparison,argno,value,jump] || [KILL,signo] || [RET,value] || [TRAP,value] || [ERRNO,errno]
```

i.e.
```
write [!=,0,1,1] [RET,0] [KILL,1]
```

meaning that for kexec_load, if argument 0 is not equal to 1, jump one past the next instruction; [if it was equal to one, then) return 0; [if we jumped past the return 0, then] kill with signal 1.

I started writing it without the brackets as I thought it woudl be nicer, but really it's kinda confusing without some sort of marker.

**Comment by lifeng68** [comment]
@hallyn 
```
What exactly should happen with:

kexec_load errno 1 args:3 [0,1,0,SCMP_CMP_LE][3,1,0,SCMP_CMP_GT][5,1,0,SCMP_CMP_MASKED_EQ]
?

errno 1 is the default action; then there are some conditions... but what happens when those conditions are met?
```
In the lxc seccomp config  version 2, the full format is like that:
```
2
blacklist allow #here is the default action
kexec_load errno 1  [0,1,0,SCMP_CMP_LE][3,1,0,SCMP_CMP_GT][5,1,0,SCMP_CMP_MASKED_EQ]
```

`kexec_load errno 1  [0,1,0,SCMP_CMP_LE][3,1,0,SCMP_CMP_GT][5,1,1,SCMP_CMP_MASKED_EQ]`
When we call `kexec_load  ` and  if those conditions are met at the same time (argument 0 is  less than or equal to 1 and argument 3 is  greater than  1 and the masked argument  5 is equal to the masked datum value 1 ), the syscall will return value of 1. Otherwise the kernel will take the default action(defined in the `blacklist allow #here is the default action`)

`valueTwo is only ever used for the masked eq, is that right?`
Yes, the valueTwo is used for masked eq

**Comment by lifeng68** [comment]
@brauner @Blub 
Thanks for your advice.
I made some improvements：
1. dropped  args:num fields
2. made `parse_v2_rule_op `  also map ==, !=, <=, etc.
3.  placed the mask value as the last and optional

**Comment by hallyn** [comment]
I see, so it's
```
[syscall] [action] [conditions-under-which-to-doaction]
```

and so I assume having multiple entries for the same syscall will be allowed?

**Comment by lifeng68** [comment]
Yes, we can give multiple entries for the same syscall, and i tested it, it works

File: src/lxc/seccomp.c [hunk 1]
```diff
@@ -129,6 +128,121 @@ static uint32_t get_and_clear_v2_action(char *line, uint32_t def_action)
...
+	else if (strcmp(s, "SCMP_CMP_GT") == 0 || strcmp(s, ">") == 0)
+		ret = SCMP_CMP_GT;
+	else if (strcmp(s, "SCMP_CMP_MASKED_EQ") == 0 || strcmp(s, "&=") == 0)
+		ret = SCMP_CMP_MASKED_EQ;
+	else
+		ret = _SCMP_CMP_MAX;
+
+	return ret;
+}
+
```
[the remark this answers is not in the document]
Comment (lifeng68) [reply src/lxc/seccomp.c]: Fixed

File: src/lxc/seccomp.c [hunk 2]
```diff
@@ -129,6 +128,121 @@ static uint32_t get_and_clear_v2_action(char *line, uint32_t def_action)
...
+		return -1;
+	}
+
+	rule_args->index = index;
+	rule_args->value = value;
+	rule_args->mask = mask;
+	rule_args->op = op;
+	return 0;
+}
+
```
[the remark this answers is not in the document]
Comment (lifeng68) [reply src/lxc/seccomp.c]: Fixed

File: src/lxc/seccomp.c [hunk 3]
```diff
@@ -129,6 +128,121 @@ static uint32_t get_and_clear_v2_action(char *line, uint32_t def_action)
...
+	char *saveptr = NULL;
+
+	tmp = strdup(line);
+	if (!tmp)
+		return -1;
+
+	/* read optional action which follows the syscall */
+	rules->action = get_v2_action(tmp, def_action);
+	if (rules->action == -1) {
+		ERROR("Failed to interpret action.");
```
[the remark this answers is not in the document]
Comment (lifeng68) [reply src/lxc/seccomp.c]: Fixed

File: src/lxc/seccomp.c [hunk 4]
```diff
@@ -129,6 +128,121 @@ static uint32_t get_and_clear_v2_action(char *line, uint32_t def_action)
...
+
+	rules->args_num = 0;
+	if (!strchr(tmp, '[')) {
+		ret = 0;
+		goto out;
+	}
+
+	for ((key = strtok_r(tmp, "]", &saveptr)), i = 0; key && i < 6; (key = strtok_r(NULL, "]", &saveptr)), i++) {
+		ret = get_seccomp_arg_value(key, &rules->args_value[i]);
+		if (ret) {
```
[the remark this answers is not in the document]
Comment (lifeng68) [reply src/lxc/seccomp.c]: Fixed

File: src/lxc/seccomp.c [hunk 5]
```diff
@@ -260,9 +374,12 @@ scmp_filter_ctx get_new_ctx(enum lxc_hostarch_t n_arch, uint32_t default_policy_
...
 
 bool do_resolve_add_rule(uint32_t arch, char *line, scmp_filter_ctx ctx,
-			uint32_t action)
+			struct seccomp_v2_rule *rule)
 {
-	int nr, ret;
+	int nr, ret, i;
+	struct scmp_arg_cmp arg_cmp[6];
+
+	memset(arg_cmp, 0x00 ,sizeof(arg_cmp));
```
[the remark this answers is not in the document]
Comment (lifeng68) [reply src/lxc/seccomp.c]: Fixed

File: src/lxc/seccomp.c [hunk 6]
```diff
@@ -129,6 +128,121 @@ static uint32_t get_and_clear_v2_action(char *line, uint32_t def_action)
...
+		if (ret) {
+			ret = -1;
+			goto out;
+		}
+		rules->args_num++;
+	}
+
+	ret = 0;
+out:
+	if (tmp)
```
[the remark this answers is not in the document]
Comment (lifeng68) [reply src/lxc/seccomp.c]: Fixed

**Comment by brauner** [comment]
jenkins: test this please

**Comment by brauner** [comment]
@lifeng68, I adapted the commit message to be a little more detailed and applied your commit otherwise unchanged in: https://github.com/lxc/lxc/commit/3ee26d1983818685643151c16496558ffcf5ced8 .
So closing here. Thanks!

**Comment by hallyn** [comment]
Thanks!

**Comment by brauner** [comment]
@hallyn, this is not necessarily set in stone. If you habe objections or want more changes, leave comments or open an issue then I can address. :)

**Pull request closed without merging**